src/output/plugins/OSXOutputPlugin.cxx

```diff
@@ -883,8 +913,11 @@ OSXOutput::Play(const void *chunk, size_t size)
 			throw std::runtime_error("Unable to restart audio output after pause");
 		}
 	}
+	if (cancel) {
+		return 0;
```

Comment:
This silently loses all audio data arriving between the `Cancel()` call and whenever the render thread happens to call `osx_render()` the next time. `Cancel()` should only discard data which arrived until that call; newly added data must not be lost.

Reply:
I assume when Play returns 0, it means at this call, no data is consumed, and next time it calls, it should output buffer at the same position, No?

Reply:
Actually, I was wrong, but your explanation doesn't make sense either; if `Play()==0`, when will the next call be?
Data isn't lost with your change; it's simply that `Play()` must never return 0. It must either consume at least one frame (and wait if necessary) or throw an exception.
But I admit that this requirement isn't documented properly; I'll add that.

Reply:
> But I admit that this requirement isn't documented properly

Oh, it is documented after all; API documentation says:

> The method blocks until at least one audio frame is consumed.

Reply:
keep it busy waiting in #598. We may change it to mutex/cond_var in the future.